### ISSUE TITLE
Samples created with invalid geometry

### ISSUE THREAD

**Issue by JimBacon** [Indicia-Team/warehouse#426]
Users can submit samples with complex geometries which are received by the warehouse as a geometry collection. A geometry collection doesn't support PostGIS functions such as ST_Intersection. This has been seen to give rise to SQL errors in the spatial index builder.

This has been fixed in the location model, see https://github.com/Indicia-Team/warehouse/blob/02f0061a1be75a4792c3d548deb1c66bd163a76a/application/models/location.php#L126

The same fix could be applied to the sample model in https://github.com/Indicia-Team/warehouse/blob/02f0061a1be75a4792c3d548deb1c66bd163a76a/application/models/sample.php#L363

**Comment by JimBacon** [comment]
An example of a received sample geometry collection is 
`GEOMETRYCOLLECTION(POLYGON((-368202.10940544 7038574.7754194,-368202.10940544 7038765.8679901,-368163.89089131 7039348.7003307,-368297.65569079 7039444.246616,-368316.76494785 7039606.6753011,-368011.01683475 7039797.7678718,-368297.65569079 7039988.8604425,-369052.471345 7040418.8187265,-369320.00094396 7040619.4659257,-369300.89168689 7040734.1214682,-369539.75740025 7040801.0038679,-369673.52219973 7040934.7686674,-369635.30368559 7041030.3149527,-369033.36208793 7040762.7853538,-368995.14357379 7040820.113125,-368890.04265991 7040772.3399823,-368899.59728845 7040743.6760967,-368507.85751854 7040495.2557548,-368297.65569079 7040294.6085556,-368039.68072036 7040084.4067278,-367896.36129234 7040132.1798705,-367686.15946459 7039826.4317574,-367915.47054941 7039644.8938152,-368039.68072036 7039616.2299296,-368116.11774863 7039578.0114155,-368106.5631201 7039520.6836443,-367982.35294915 7039167.1623885,-367915.47054941 7039071.6161032,-368097.00849156 7038622.5485621,-368202.10940544 7038574.7754194)),LINESTRING(-368202.10940544 7038555.6661623,-368202.10940544 7038574.7754194))`

**Comment by johnvanbreda** [comment]
@JimBacon I agree with the proposal, with the caveat that I'd like to know there isn't a significant performance hit of the processing code on saving a sample. We save many more samples than locations.

**Comment by JimBacon** [comment]
The time to clean up a geometry will be part initiating the query and part executing it. The time to execute will depend on the complexity of the geometry. All will depend on the server load at the time of execution.

Using the geometry collection in the above comment as one case and `POINT(-69898.6214640035 6678609.19780742)` as another, plugging them in to the query 

```
SELECT CASE
  WHEN ST_GeometryType(ST_CollectionHomogenize(ST_MakeValid(ST_GeomFromText('$value', $srid)))) = 'ST_GeometryCollection' THEN
    ST_Buffer(ST_MakeValid(ST_GeomFromText('$value', $srid)), 0.00001, 'quad_segs=2')
  ELSE
    ST_CollectionHomogenize(ST_MakeValid(ST_GeomFromText('$value', $srid)))
END AS geom
```

where `$srid` is replaced by `9000913` and `$value` by the WKT resulted in execution times in the following ranges over a dozen attempts at each using pgAdmin.

- With Point geom, 87 - 142ms with one outlier at 1623ms
- With Collection geom, 86 - 187ms

Both were typically about 100ms.

Performing the same cleanup on a random selection of 10000 sample geoms (as WKB or WKT) took 2 - 5 seconds.

The indication is, therefore, that the execution time is negligible compared to the set up time whether a collection or not.

In my test, pgAdmin was connecting to the database over a network whereas, in our current implementation, the warehouse code runs on the same machine as the database so I would expect a faster set up time than I experienced.
